docs/src/main/mdoc/datatypes/statet.md

```diff
@@ -0,0 +1,285 @@
+  state machine.
+- `A` represents the return value.
+
+It can be seen as a way to add the capability to manipulate a shared
+state to an existing computation in the context of `F`.
+
+This definition could be confusing, but it will become clear after
+learning the `State` monad and by the example below.
+
+## StateT and State Relationship
+
+`StateT` is a monad transformer for
+[State](https://typelevel.org/cats/datatypes/state.html), in
+particular [State](https://typelevel.org/cats/datatypes/state.html) is
+defined as a `StateT` with
+[Eval](https://typelevel.org/cats/datatypes/eval.html) as the effect
+`F`.
+
+```scala mdoc:silent
+import cats.data.StateT
+import cats.Eval
+
+type State[S, A] = StateT[Eval, S, A]
+```
+
+Therefore, `StateT` exposes the same methods of
+[State](https://typelevel.org/cats/datatypes/state.html), such as:
+`modify`, `get` and `set`.  Plus additional methods, that handles
+effectful computations, eg: `modifyF`, `setF` and `liftF`.
+
+## Example: Table Reservation System
+
+In this example, we are going to model a *Table Reservation System*. To
+do so, we need to keep track of the tables and the current
+reservations for each of them. The end-user can then try to insert a
+booking for a specific table and time. If such a table is available,
+then the booking is placed and the state is updated, otherwise, an
+error is returned.
+
+To simplify the logic, for each reservation we will just consider a
+single `LocalTime` starting at the beginning of the hour.
+
+Let's start with defining the type synonym for the effect type:
+
+```scala mdoc:silent:reset
+import cats.data.{StateT, NonEmptyList}
+import cats.implicits._
+import java.time.LocalTime
+
+type ThrowableOr[A] = Either[Throwable, A]
+```
+
+We will use an `Either[Throwable, A]` because each
+booking insertion can fail.
+
+Now, we need to implement/define:
+- The type representing the reservation.
+- The type representing the state of the Table Reservation System. It
+  will wrap around a collection of Reservations.
+- An initial state, that will be just an empty state type.
+- A custom `Throwable` to be used in case of an error.
+- The logic for the booking insertion. We can take advantage of the
+method `modifyF` later on to apply it to the system state.
+
+In addition, we can implement a simple function that will evaluate a
+`NonEmptyList` of reservations, inserting them one by one.
+
+```scala mdoc:silent
+object TableReservationSystem {
+
+  final case class ReservationId(tableNumber: Int, hour: LocalTime)
+  final case class Reservation(id: ReservationId, name: String)
+  final case class Reservations(reservations: List[Reservation]) {
+    def insert(reservation: Reservation): ThrowableOr[Reservations] =
+      if (reservations.exists(r => r.id == reservation.id))
+        Left(TableAlreadyReserved(reservation))
+      else Right(Reservations(reservations :+ reservation))
+  }
+
+  final case class TableAlreadyReserved(
```

Comment:
```suggestion
  final class TableAlreadyReservedException(
```

A personal thing but I think using `case class` for exceptions is an anti-pattern.